src/contract/Signature.ts

```diff
@@ -29,8 +36,62 @@ export class Signature {
       this.assertEnvForCustomSigner(walletOrSignature);
       this.signer = walletOrSignature.signer;
       this.type = walletOrSignature.type;
+      this.signatureProviderType = 'CustomSignature';
+    } else if (this.isBundlerSignature(walletOrSignature)) {
+      this.signatureProviderType = 'BundlerSigner';
+      this.type = decodeBundleSignatureType(walletOrSignature.signatureType);
     } else {
-      this.assignDefaultSigner(walletOrSignature);
+      this.assignArweaveSigner(walletOrSignature);
+      this.signatureProviderType = 'ArWallet';
+      this.type = 'arweave';
+    }
+    this.wallet = walletOrSignature;
+  }
+
+  async getAddress(): Promise<string> {
+    if (this.cachedAddress) {
+      return this.cachedAddress;
+    }
+
+    switch (this.signatureProviderType) {
+      case 'CustomSignature': {
+        if (this.wallet.getAddress) {
+          this.cachedAddress = await this.wallet.getAddress();
+        } else {
+          this.cachedAddress = await this.deduceSignerBySigning();
+        }
+        return this.cachedAddress;
+      }
+      case 'ArWallet': {
+        this.cachedAddress = await this.deduceSignerBySigning();
+        return this.cachedAddress;
+      }
+      case 'BundlerSigner': {
+        // TODO: this can be done without additional signing step if we can be sure that .publicKey is always set in bundlr impl
```

Comment:
aren't we sure? even in `isBundlerSignature` method we are checking for this field.

Reply:
True we shouldn't be checking for this field, will fix cause: https://github.com/Bundlr-Network/arbundles/blob/9fafdbfec6fbfcbcb538b92ae9bd0d9fbe413fb8/src/signing/chains/injectedEthereumSigner.ts#L17 we can;t be sure that user will call 'setPublicKey' (and it will generate additional pop up from metamask anyway)

Also we would have to convert  pub key  to ETH address so some dependency might be needed.
Last but not least, we don't use BundlerSigner for interactions yet, so  this method is not used in sdk in context of writing interaction.

Reply:
yeah, I mean having signer without publicKey does not make much sense as we cannot sign the data item with arbundles. I was just thinking if we should allow creating Signature without it and should we maybe check and validate it.

The problem only occurs in browser environment, for server public key is assigned automatically 
1. https://github.com/Bundlr-Network/arbundles/blob/master/src/signing/chains/ArweaveSigner.ts#L11
2. https://github.com/Bundlr-Network/arbundles/blob/master/src/signing/keys/Rsa4096Pss.ts#L16

Anyway - we can leave it for now and return to the subject while implementing it for interactions.

Reply:
Handled partially, rest will be implemented with bundling interactions :)